### PR TITLE
Submission by Sebastian Scheibe

### PR DESCRIPTION
**Pull request by Ecostack**
This is the submission for the Wiredcraft Test Backend assignment.

Please see the README.md for more information on the project can be started.

It can be seen live on this URL: https://wiredcraft.techstack.de/
Best to start with the API documentation: https://wiredcraft.techstack.de/api

This is still work in progress for some things, so please bear with me.

### CODE REVIEW

**Comment by Ecostack** [comment]
I will put the PR into ready for review, although I will update the e2e tests and will add further code comments.

File: README.md [hunk 1]
```diff
@@ -1,73 +1,57 @@
...
-- Imagine we have a new requirement right now that the user instances need to link to each other, i.e., a list of "followers/following" or "friends". Can you find out how you would design the model structure and what API you would build for querying or modifying it?
-- Related to the requirement above, suppose the address of user now includes a geographic coordinate(i.e., latitude and longitude), can you build an API that,
-  - given a user name
-  - return the nearby friends
-
+## Architecture
 
-## What We Care About
+This project uses the onion architecture which allows for maintainability.
+Please check out the `docs/architecture.md` for further informations.
```
Comment (xavierchow) [inline README.md:41]: ah, I found it should be `documentation/architecture.md`

File: src/presentation/rest/user/user.controller.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+  @Post()
+  create(@Body() createUserDto: CreateUserDto) {
+    return this.userService
+      .create({
+        ...createUserDto,
+        address: this.transformLatLongToPostgresGeoPosition(
+          createUserDto.address,
+        ),
+      })
+      .then((res) => this.mapUserAddressToOutside(res));
```
Comment (xavierchow) [inline src/presentation/rest/user/user.controller.ts:34]: Isn't it better to use pipes / interceptors for those transformations?

File: README.md [hunk 3]
```diff
@@ -1,73 +1,57 @@
...
-  "id": "xxx",                  // user ID 
-  "name": "test",               // user name
-  "dob": "",                    // date of birth
-  "address": "",                // user address
-  "description": "",            // user description
-  "createdAt": ""               // user created date
-}
+POST    /user
+GET     /user
+GET     /usr/{id}
```
[the remark this answers is not in the document]
Comment (Ecostack) [reply README.md]: Indeed user, corrected the typo

File: src/presentation/rest/user/user.controller.spec.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+      findAllUsers = await controller.findAll();
+      expect(findAllUsers).toHaveLength(0);
+
+      const foundUser = await controller.findOne(createdUser.id);
+      expect(foundUser).toEqual(undefined);
+    });
+
+    it('Add 1 user with full details', async () => {
+      await controller.create({
```
[the remark this answers is not in the document]
Comment (Ecostack) [reply src/presentation/rest/user/user.controller.spec.ts]: Fixed that, does assert now

File: src/infrastructure/in-memory/user/user.repository.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  async deleteById(params: { id: string }): Promise<boolean> {
+    if (this.users.has(params.id)) {
+      this.users.delete(params.id);
+      return true;
+    }
+    return false;
+  }
+
+  async findAll(params: { offset: number; limit: number }): Promise<User[]> {
+    return [...this.users.values()];
```
[the remark this answers is not in the document]
Comment (Ecostack) [reply src/infrastructure/in-memory/user/user.repository.ts]: Fixed that one too, now being tested

File: README.md (hunk 1 above)
Comment (Ecostack) [reply README.md]: Fixed the typo for the path name docs -> documentation

File: test/app.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    const user = await createUser({
+      name: 'Philip J. Fry 1st',
+    });
+
+    let returnedUser = await getUser(user.id);
+    expect(returnedUser).toEqual(user);
+    await deleteUser(user.id);
+
+    returnedUser = await getUser(user.id);
+    expect(returnedUser).toStrictEqual({});
```
[the remark this answers is not in the document]
Comment (Ecostack) [reply test/app.e2e-spec.ts]: Indeed, returns now 404

File: src/presentation/rest/user/user.controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+      expect(await controller.remove(createdUser.id)).toEqual(false);
+
+      findAllUsers = await controller.findAll();
+      expect(findAllUsers).toHaveLength(0);
+
+      const foundUser = await controller.findOne(createdUser.id);
+      expect(foundUser).toEqual(undefined);
+    });
+
+    it('Add 1 user - delete should remove user from get and findAll', async () => {
```
[the remark this answers is not in the document]
Comment (Ecostack) [reply src/presentation/rest/user/user.controller.spec.ts]: Fixed that mistake

File: src/presentation/rest/user/user.controller.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    private readonly userService: UserService,
+    private readonly friendService: FriendService,
+  ) {}
+
+  @Post()
+  create(@Body() createUserDto: CreateUserDto) {
+    return this.userService
+      .create({
+        ...createUserDto,
+        address: this.transformLatLongToPostgresGeoPosition(
```
[the remark this answers is not in the document]
Comment (Ecostack) [reply src/presentation/rest/user/user.controller.ts:31]: I think the naming there is a bit wrong, it the end, the data model is independent of PostgreSQL.

File: src/presentation/rest/user/user.controller.ts (hunk 2 above)
Comment (Ecostack) [reply src/presentation/rest/user/user.controller.ts:35]: Indeed, thanks for the heads up. Gonna rewrite that logic using interceptors.

Comment (xavierchow) [reply src/presentation/rest/user/user.controller.ts:35]: Thanks, just a reminder it's not necessary to put a fix or improvement for those comments.
After all, it's a test, we are not expecting a product-ready code here. :P

Comment (Ecostack) [reply src/presentation/rest/user/user.controller.ts]: Fair enough, I just took the opportunity to finish it :)

**Comment by JuhaS** [comment]
@Ecostack I would like to see if how you would improve the friend follow implementation in terms of making it cleaner solution (not adding more features), related to friend.repository.ts.

**Comment by Ecostack** [comment]
I assume this is regarding the PostgreSQL implementation and its way of maintaining the relationship between users.

I see potential in removing the OR clause in case both user ids are available. If the order of user ids is always the same, for example by sorting them, the userId and otherUserId will always be the same for a record. With that in place, the CREATE, FIND and DELETE operation, which involve two ids, can be made easier. 

It would be like:

User (id: 546) creates friendship with user (id: 325). Ids sorted ascending, the friendship record is created with userId (325) and otherUserId (546). Same order would be applied if user id (325) creates/finds/deletes the record.

Further on, the status of a friendship only allows created and deleted. It would be great to add a status attribute to it.

**Pull request closed without merging**